dev/requirements.txt

```diff
@@ -1,5 +1,8 @@
-flake8==3.5.0
+pycodestyle==2.5.0
```

Comment:
By the way @HyukjinKwon, following on from the discussion here https://github.com/apache/spark/pull/27534#discussion_r377394628, this is a good example of why it's good to pin versions of build/test dependencies. This release of pycodestyle adds new checks that we did not ignore in our `tox.ini` config.

So if this had been running without a pinned version, one day our build would have start failing because pycodestyle made a new release with new checks and our build automatically started using it.

Reply:
On the other hand, not pinning will trigger us to fix quickly to support latest versions. Pinning version will force users to install different versions of libraries they already have. If something is found we can't easily fix, we could pin the version. If I am not completely wrong, using the latest version by default is the conventional approach in Python side (and R side - it reminds me of [`testthat` case](https://github.com/apache/spark/pull/27359)). This is the approach used in [pandas as an example](https://github.com/pandas-dev/pandas/blob/master/requirements-dev.txt). [I used this approach too](https://github.com/databricks/koalas/blob/master/requirements-dev.txt).

I agree, in a way, it can make more sense to pin the version in particular when this is used for CI purpose or release docker image which are supposed to discard of after using once; however, I don't quite agree when this affects other developers' environments.

cc @dongjoon-hyun. IIRC, I think we have related discussions at https://github.com/apache/spark/pull/26824.

Reply:
I always work on Spark from within a dedicated Python virtual environment, so any requirements specified there do not affect my work on any other project. I work with independent Python virtual environments as a matter of habit; perhaps that's why I don't consider it a disruption to developer environments to pin versions.

As for what's common convention in the Python community, there is a shift among some high profile projects towards using tools like pip-tools, pipenv, or poetry, to completely specify and pin all direct and transitive dependencies.

Here are a couple of examples that use pip-sync, for example:
* Warehouse (the software behind PyPI): https://github.com/pypa/warehouse/blob/master/requirements/tests.txt
* Trio (next gen Python async library): https://github.com/python-trio/trio/blob/master/test-requirements.txt